### ISSUE TITLE
Issues trying to use ESPv2 with gRPC-web with CORS

### ISSUE THREAD

**Issue by Despicable-bee** [GoogleCloudPlatform/esp-v2#192]
I've recently been trying to ping my backend from the browser using the client stub generated from the following commands:

Generate client stubs
```
protoc -I=. \
    --proto_path=. \
    --js_out=import_style=commonjs:. example.proto 

protoc -I=. \
    --grpc-web_out=import_style=commonjs,mode=grpcwebtext:. example.proto
```

Client code
```
import { Injectable } from '@angular/core';

import { ApplicationProcessorClient } from './stub/pluto-website_grpc_web_pb';
import { DetailsRequest, DetailsResponse } from './stub/pluto-website_pb';

@Injectable({
  providedIn: 'root'
})
export class ApplicationClientService {

  constructor() {
    this.ApplicationService = new ApplicationProcessorClient('https://ESPv2_container-abcdefg-an.a.run.app');
  }

public send_details() {
let request = new DetailsRequest()
    request.setTitle("Mr");
    request.setFirstName("firstName");
    request.setLastName("lastName");
   
    const metadata = {};
    this.ApplicationService.confirmDetails(request, metadata, function(err, response) {
      console.log(response);
      console.log(err);
    })
}


private ApplicationService;
}
```

I've tested my server code with a separate python script and it works as expected.

I've deployed my ESPv2 container with the following commands (in order)

Update the endpoints definition
```
gcloud endpoints services \
    deploy api_descriptor.pb \
    api_config.yaml
```

Build the image
```
./gcloud_build_image -s ESPv2_container-abcdefg-an.a.run.app \
        -c 2020-xx-xxr0 \
        -p gcloud-project-id
```

Deploy the image
```
gcloud run deploy espv2-container \
    --image="gcr.io/gcloud-project-id/endpoints-runtime-serverless:ESPv2_container-abcdefg-an.a.run.app-2020-xx-xxr0"\
    --set-env-vars=ESPv2_ARGS=--cors_preset=basic,--cors_allow_headers="User-Agent"\
    --allow-unauthenticated --platform managed\
    --project gcloud-project-id
```

When I call the client code from the browser, I get the following error

```
Access to XMLHttpRequest at 'https://ESPv2_container-abcdefg-an.a.run.app/endpoints.application.ApplicationProcessor/ConfirmDetails' from origin 'http://localhost:4200' has been blocked by CORS policy: Request header field x-user-agent is not allowed by Access-Control-Allow-Headers in preflight response.
```

I'm not quite sure why this is the case. The google [docs](https://cloud.google.com/endpoints/docs/grpc/specify-esp-v2-startup-options#cors) specify that the `--cors=basic` config should include `User-Agent`. I tried adding the flag manually to the response header with the `--cors_allow_headers="User-Agent'` option, doesn't make a difference.

I'm really stumped as to what's gone wrong, I have a basic understanding of CORS, so if its something dumb can someone please point it out for me?

Kind regards,

Despicable-b

**Comment by nareddyt** [comment]
As a sidenote, I noticed `ESPv2_ARGS` in your deploy command is mis-formatted. There is a specific syntax you need to follow if you need to pass more than one arg (due to a limitation in the `gcloud run deploy` command). 

[Reference doc](https://cloud.google.com/endpoints/docs/grpc/specify-esp-v2-startup-options#setting-configuration-flags) have and explanation and example:

```
--set-env-vars=ESPv2_ARGS=^++^--cors_preset=basic++--cors_allow_origin=your_host.com
```

If you deployed with only the `--cors_preset=basic` option, this should have worked. Still looking into it.

**Comment by nareddyt** [comment]
The error you posted mentions `x-user-agent` is not valid. Not sure where the `x-` prefix is coming from, but `User-Agent` is the [standard header name](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent). This seems like a custom header you set different from `user-agent`.

Maybe you can try to:
- Check if your code is sending `x-user-agent` instead of `user-agent`? Modify it to send the correct one?
- Follow the comment above to specify `--cors_allow_headers` with `x-user-agent`?

FYI you manually specify `--cors_allow_headers`, you will overwrite the default headers that are configured by `--cors_preset=basic`. You can copy all the [values here](https://github.com/GoogleCloudPlatform/esp-v2/blob/2f523997feb29a95639db856865c9404c5d38730/docker/generic/start_proxy.py#L276) and append `x-user-agent` if needed.

**Comment by Despicable-bee** [comment]
I tried running the following command to deploy the container to the cloud:

```
gcloud run deploy espv2-container-name \
    --image="gcr.io/project-id/endpoints-runtime-serverless:espv2-container-name-abcdefg-an.a.run.app-${1}"\
    --set-env-vars=ESPv2_ARGS=^++^--cors_preset=basic++--cors_allow_headers="DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization"\
    --allow-unauthenticated --platform managed\
    --project project-id
```

 but the gcloud cli tool keeps giving me this error:
```
(gcloud.run.deploy) argument --set-env-vars: Bad syntax for dict arg: [User-Agent]. Please see `gcloud topic flags-file` or `gcloud topic escaping` for information on providing list or dictionary flag values with special characters.
```

I followed the specification outlined in the [reference doc](https://cloud.google.com/endpoints/docs/grpc/specify-esp-v2-startup-options#cors), so I'm unsure why I'm getting this error.

Any ideas?

**Comment by nareddyt** [comment]
I see, so the grpc-web client library sets `x-user-agent` and `x-grpc-web` headers. Then you can't workaround that, it's built into the client.

I'm not sure why the last comment you made did not work. If you manually allow the headers, it should have accepted them. I am currently out of ideas. I'll try to debug some more, but we can also wait until next week and have @qiwzhang take a look.

We did have an issue with grpc-web a few days ago. It was fixed but not released yet (#176). However, I don't think this is related, as your browser is not even sending the actual request to the server due to the preflight response.

**Comment by qiwzhang** [comment]
x-grpc-web header is required. could you add it to --cors-allow_headers?

**Comment by Despicable-bee** [comment]
I've added the following

```
BASE_IMAGE=gcr.io/endpoints-release/endpoints-runtime-serverless:2

function error_exit() {
  # ${BASH_SOURCE[1]} is the file name of the caller.
  echo "${BASH_SOURCE[1]}: line ${BASH_LINENO[0]}: ${1:-Unknown Error.} (exit ${2:-1})" 1>&2
  exit ${2:-1}
}

while getopts :c:s:p:i: arg; do
  case ${arg} in
    c) CONFIG_ID="${OPTARG}";;
    s) SERVICE="${OPTARG}";;
    p) PROJECT="${OPTARG}";;
    i) BASE_IMAGE="${OPTARG}";;
    \?) error_exit "Unrecognized argument -${OPTARG}";;
  esac
done

[[ -n "${PROJECT}" ]] || error_exit "Missing required PROJECT"
[[ -n "${SERVICE}" ]] || error_exit "Missing required SERVICE"
[[ -n "${CONFIG_ID}" ]] || error_exit "Missing required CONFIG_ID"
echo "Using base image: ${BASE_IMAGE}"

cd "$(mktemp -d /tmp/docker.XXXX)"

# Be careful about exposing the access token.
curl --fail -o "service.json" -H "Authorization: Bearer $(gcloud auth print-access-token)" \
  "https://servicemanagement.googleapis.com/v1/services/${SERVICE}/configs/${CONFIG_ID}?view=FULL" \
  || error_exit "Failed to download service config"

(
set -x

cat <<EOF > Dockerfile
FROM ${BASE_IMAGE}
USER root
ENV ENDPOINTS_SERVICE_PATH /etc/endpoints/service.json
COPY service.json \${ENDPOINTS_SERVICE_PATH}

ENV ESPv2_ARGS ^++^--cors_preset=basic++--cors_allow_headers="DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,X-Grpc-Web,X-User-Agent"

RUN chown -R envoy:envoy \${ENDPOINTS_SERVICE_PATH} && chmod -R 755 \${ENDPOINTS_SERVICE_PATH}
USER envoy
ENTRYPOINT ["/env_start_proxy.py"]
EOF

NEW_IMAGE="gcr.io/${PROJECT}/endpoints-runtime-serverless:${SERVICE}-${CONFIG_ID}"
gcloud builds submit --tag "${NEW_IMAGE}" . --project="${PROJECT}"
)
```
However I still get the same error asking for the X-User-Agent header

Interestingly enough you can take a look at the yaml on a deployed container. For some reason, it only contains the --cors_preset=basic flag. Even after manually editting, and including the 
```
++--cors_allow_headers="DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,X-Grpc-Web,X-User-Agent"
```
flag, it still doesn't work.

My deployed yaml file looks like the following

```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: espv2-name
  namespace: '146637902517'
  selfLink: /apis/serving.knative.dev/v1/namespaces/146637902517/services/espv2-name
  uid: 1234567890abcdefghi
  resourceVersion: ABCDEFGHI
  generation: 36
  creationTimestamp: '2020-05-17T13:49:34.922795Z'
  labels:
    cloud.googleapis.com/location: asia-northeast1
  annotations:
    run.googleapis.com/client-name: cloud-console
    serving.knative.dev/creator: none.of.your.business@gmail.com
    serving.knative.dev/lastModifier: none.of.your.business@gmail.com
    client.knative.dev/user-image: gcr.io/project-id/endpoints-runtime-serverless:espv2-name-abcdefgh-an.a.run.app-2020-06-15r0
spec:
  traffic:
  - percent: 100
    latestRevision: true
  template:
    metadata:
      name: espv2-name-00035-mus
      annotations:
        run.googleapis.com/client-name: cloud-console
        client.knative.dev/user-image: gcr.io/project-id/endpoints-runtime-serverless:espv2-name-abcdefgh-an.a.run.app-2020-06-15r0
        autoscaling.knative.dev/maxScale: '1000'
    spec:
      timeoutSeconds: 900
      serviceAccountName: 123456789-compute@developer.gserviceaccount.com
      containerConcurrency: 80
      containers:
      - image: gcr.io/project-id/endpoints-runtime-serverless:espv2-name-abcdefgh-an.a.run.app-2020-06-15r0
        ports:
        - containerPort: 8080
        env:
        - name: ESPv2_ARGS
          value: ^++^--cors_preset=basic++--cors_allow_headers="DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,X-Grpc-Web,X-User-Agent"
        resources:
          limits:
            cpu: 1000m
            memory: 244Mi
status:
  conditions:
  - type: Ready
    status: 'True'
    lastTransitionTime: '2020-06-15T05:06:21.790Z'
  - type: ConfigurationsReady
    status: 'True'
    lastTransitionTime: '2020-06-15T05:06:15.881Z'
  - type: RoutesReady
    status: 'True'
    lastTransitionTime: '2020-06-15T05:06:21.790Z'
  observedGeneration: 36
  traffic:
  - revisionName: espv2-name-00035-mus
    percent: 100
    latestRevision: true
  latestReadyRevisionName: espv2-name-00035-mus
  latestCreatedRevisionName: espv2-name-00035-mus
  address:
    url: https://espv2-name-abcdefgh-an.a.run.app
  url: https://espv2-name-abcdefgh-an.a.run.app
```
Can you check that this happens when deploying one of your own containers and trying to ping some method using the following generated stubs?

```
protoc -I=. \
    --proto_path=. \
    --js_out=import_style=commonjs:. your-methods.proto 

protoc -I=. \
    --grpc-web_out=import_style=commonjs,mode=grpcwebtext:. your-methods.proto
```

Kind regards,

Despicable-bee

**Comment by qiwzhang** [comment]
I run into this grpc-web envoy example [config](https://github.com/grpc/grpc-web/blob/master/net/grpc/gateway/examples/helloworld/envoy.yaml)

Could you try following cors flags

ENV ESPv2_ARGS ^++^--cors_preset=basic++--cors_allow_headers="keep-alive,user-agent,cache-control,content-type,content-transfer-encoding,custom-header-1,x-accept-content-transfer-encoding,x-accept-response-streaming,x-user-agent,x-grpc-web,grpc-timeout"++--cors_expose_headers="grpc-status,grpc-message"

**Comment by Despicable-bee** [comment]
Yup that did the trick, although when initially running the build and deploy commands for the container, it did not update the environment variable.

Perhaps this is something wrong on my end but I did follow the same steps as outlined previously.

Regardless I was able to edit the yaml using the cloud console and all went smoothly (got the lovely HTTP 200 response).

Thank you for your diligence and support, I really appreciate it.

Kind regards,

Despicable bee.
